Feature inv plugin conf

chrismeyersfsu opened the pull request:
related to https://github.com/ansible/awx/issues/5150

### TODO:
- [x] After talking with @AlanCoding . I have `source="scm"` all wrong. `source_vars` are irrelevant  for SCM source & we should be pointing to the inventory file in the SCM, which will point to the inventory plugin via `plugin: `
- [x] Remove the migration rewind before merge.

chrismeyersfsu committed:
migrate to new style inv plugin

chrismeyersfsu committed:
remove group_by from inventory source

* Does not remove group_by testing

chrismeyersfsu committed:
remove instance_filter

chrismeyersfsu committed:
remove source_regions

chrismeyersfsu committed:
fix tests for new inv plugin behavior

* Enforce plugin:

chrismeyersfsu commented on awx/main/models/inventory.py:
```diff
@@ -2028,76 +1467,10 @@ def get_plugin_env(self, *args, **kwargs):
     def inventory_as_dict(self, inventory_update, private_data_dir):
-        ret = super(gce, self).inventory_as_dict(inventory_update, private_data_dir)
+        ret = super().inventory_as_dict(inventory_update, private_data_dir)
         credential = inventory_update.get_cloud_credential()
```
This is the 1 inventory plugin type where the config lives in a credential and has to be transferred to the inventory `gce.yml`. We can solve this in the inventory plugin by accepting the project list as an env var & update our credential type accordingly.

chrismeyersfsu committed:
forgot to add migration helper

* move models/inventory.py plugin injector logic to a place frozen in
time useable by the migration code.

chrismeyersfsu commented on awx/main/migrations/0118_v380_inventory_plugins.py:
```diff
@@ -0,0 +1,83 @@
+            inv_source_obj.save()
+
+
+def inventory_source_vars_backward(apps, schema_editor):
```
I plan on removing this before merge. It's nice to be able to backward when developing but alas customers almost certainly cram sensitive info into their source_vars

chrismeyersfsu commented on awx/main/models/inventory.py:
```diff
@@ -1419,6 +1158,20 @@ def clean_source_path(self):
+        injector = self.injectors.get(self.source)
+        source_vars = dict(self.source_vars_dict) # make a copy
+        if injector and self.source_vars_dict.get('plugin', '') != injector.get_proper_name():
+            source_vars['plugin'] = injector.get_proper_name()
```
Add or correct `plugin: a.b.c` for user based on `source` property.

chrismeyersfsu commented on awx/main/models/inventory.py:
```diff
@@ -1419,6 +1158,20 @@ def clean_source_path(self):
+        source_vars = dict(self.source_vars_dict) # make a copy
+        if injector and self.source_vars_dict.get('plugin', '') != injector.get_proper_name():
+            source_vars['plugin'] = injector.get_proper_name()
+        elif not injector:
```
`scm` source type requires the user to specify `plugin: a.b.c`, we can't do it for them. Note that this _could_ break some flows because we don't allow the model to be created/modified if the structure of their `source_vars` isn't correct. We _could_ instead allow `plugin: ` to not be present and fail at runtime.

AlanCoding commented:
What UI work would be needed to wrap this up?

AlanCoding commented on awx/main/migrations/0118_v380_inventory_plugins.py:
```diff
@@ -0,0 +1,83 @@
+
+
+def _get_inventory_sources(InventorySource):
+    # TODO: Maybe pull the list of cloud sources from code
```
I don't know, it's probably better to use a frozen list for the migration

chrismeyersfsu commented:
Re: UI work needed. Nothing new.
![image](https://user-images.githubusercontent.com/722880/88843639-77f13580-d1af-11ea-9547-9c605bb11be1.png)
Remove instance filters, regions, only group by.

chrismeyersfsu committed:
do not enforce plugin: for source=scm

* InventorySource w/ source type scm point to an inventory file via
source_file. source_vars are ignored.

jakemcdermott committed:
Delete inventory source fields

wenottingham commented:
Thinking of this from a user perspective, this trades in reduced complexity, and increased flexibility, at the expense of usability niceness.

We probably need an option to generate what the frozen migration would generate from a blank config so that users can make new sources that match what a migrated source would be.

Why do <source>_ENABLED_VAR/_ENABLED_VALUE/_GROUP_FILTER still exist in *settings*? If we're doing that much change, we should fix that so it's per-source as well.

jakemcdermott committed:
Delete inventory source fields

chrismeyersfsu committed:
safer migrations

chrismeyersfsu committed:
rename inventory migration

jakemcdermott committed:
Add new common inventory source fields

chrismeyersfsu committed:
simplify InventorySourceOption inheritance hack

jakemcdermott committed:
Add new common inventory source fields

jakemcdermott commented on awx/ui/client/src/inventories-hosts/inventories/related/sources/sources.form.js:
```diff
@@ -126,46 +126,6 @@ export default ['NotificationsList', 'i18n', function(NotificationsList, i18n){
                 includeInventoryFileNotFoundError: true,
                 subForm: 'sourceSubForm'
             },
```
![Screenshot from 2020-08-04 17-51-47](https://user-images.githubusercontent.com/9753817/89348907-5692bc80-d67b-11ea-91c2-02f557cabc65.png)

jakemcdermott commented on awx/ui_next/src/screens/Inventory/shared/InventorySourceForm.jsx:
```diff
@@ -200,21 +200,21 @@ const InventorySourceForm = ({
     verbosity: source?.verbosity || 1,
+    enabled_var: source?.enabled_var || '',
+    enabled_value: source?.enabled_value || '',
+    host_filter: source?.host_filter || '',
```
![Screenshot from 2020-08-04 17-49-01](https://user-images.githubusercontent.com/9753817/89348959-6c07e680-d67b-11ea-9013-c3f4235ceadc.png)

chrismeyersfsu committed:
fix missed import

chrismeyersfsu committed:
overwrite `plugin:` at runtime

* Before, we were re-writing `plugin:` when users updated the
InventorySource via the API. Now, we just override at run-time. This
makes for a more sane API interaction

chrismeyersfsu committed:
inventory plugin inventory parameter take precedence over env vars

chrismeyersfsu committed:
push global invsource fields onto invsource obj

chrismeyersfsu commented on awx/main/migrations/0118_inventory_plugins.py:
```diff
@@ -0,0 +1,93 @@
+            inv_source_obj.save()
+
+
+def inventory_source_vars_backward(apps, schema_editor):
```
TODO: Remove the rewind. Users could have sensitive data.

AlanCoding commented on awx/ui/client/src/inventories-hosts/inventories/related/sources/sources.form.js:
```diff
@@ -340,6 +300,36 @@ export default ['NotificationsList', 'i18n', function(NotificationsList, i18n){
+                type: 'text',
+                dataTitle: i18n._('Host Filter'),
+                dataPlacement: 'right',
+                awPopOver: "<p>" + i18n._("Regular expression where only matching hosts will be imported.") + "</p>",
```
I think it would be good to mention that this is client-side filtering.

I'm still very worried about confusion with these fields, because there is duplication between them and the inventory plugin options themselves. I wouldn't even hesitate saying that "this is not an inventory plugin parameter"

jakemcdermott replied:
updated:
![Screenshot from 2020-08-27 12-18-48](https://user-images.githubusercontent.com/9753817/91468222-ad954700-e85f-11ea-931d-3bcb68c76f08.png)
![Screenshot from 2020-08-27 12-19-01](https://user-images.githubusercontent.com/9753817/91468232-b0903780-e85f-11ea-9867-48d3daac2783.png)
![Screenshot from 2020-08-27 12-19-13](https://user-images.githubusercontent.com/9753817/91468241-b38b2800-e85f-11ea-976c-75bba2b232cf.png)

cc @dsesami

AlanCoding commented:
I'm pulling this down and trying to test some things

![Screen Shot 2020-08-21 at 10 44 00 AM](https://user-images.githubusercontent.com/1385596/90902986-4d079500-e39b-11ea-92b2-2c1e7c67f00a.png)

I think this should be a little more explicit in order to avoid the inevitable confusion. The Source Variables (can we please call this something else?) may still have filtering options. But this performs an additional filter after the host list is obtained with regex filtering. Also, does anyone know how to write these? Does it filter out hosts if it matches a given expression, or does it only _include_ hosts that match the expression? If I give multiple terms... are they an OR condition or AND?

elyezer commented:
I am exploring this and found that the value of `source_vars` returned by the API is not consistent. For example, custom source returns the `source_vars` as being a JSON string but others are returning a YAML string. That makes me thing if the source vars should be normalized to JSON (what we had previously) for everything.

chrismeyersfsu commented:
`source_vars` is a TextField and is not coerced into JSON or YAML when stored because there was a want to store comments. So the behavior you are seeing is expected.

The migration in this PR is the exception. We coerce the `source_vars` to yaml.

elyezer commented:
@chrismeyersfsu thanks for the clarification, wanted to make sure we were not missing anything.

dsesami reviewed (changes requested):
This looks great from a UI standpoint, but I'd like to request a small adjustment:
`1.` Require both enabled var and value and throw an error if only one is set. This is not what the API does, and I assume (perhaps incorrectly) that there is a reason for that.
`2.` If we want to keep validation as-is, can we update the tooltip on enabled_value to mention that it is only used if enabled_var is set, and vice-versa?

@jakemcdermott have agreed to do `#1` and `#2` for ui_next and `#2` for the classic ui.

bonus points: can we have the client double-check that the regex syntax for the host filter is valid? small quality of life change

jakemcdermott committed:
Add details to inv source field help text

jakemcdermott committed:
Check that host_filter is regexp

jakemcdermott commented:
Alright, this is ready for more review and testing. Thanks for taking a look

jakemcdermott commented on awx/ui_next/src/screens/Inventory/shared/InventorySourceSubForms/SharedFields.jsx:
```diff
@@ -192,6 +192,7 @@ export const HostFilterField = withI18n()(({ i18n }) => {
       name="host_filter"
       type="text"
+      validate={regExp(i18n)}
     />
```
![Screenshot from 2020-08-27 11-59-49](https://user-images.githubusercontent.com/9753817/91469005-cfdb9480-e860-11ea-93df-8923f01ec65a.png)
cc @dsesami

jakemcdermott commented on awx/ui/client/src/inventories-hosts/inventories/related/sources/sources.form.js:
```diff
@@ -340,6 +300,36 @@ export default ['NotificationsList', 'i18n', function(NotificationsList, i18n){
+                dataContainer: 'body',
+                ngDisabled: '!(inventory_source_obj.summary_fields.user_capabilities.edit || canAdd)',
+                subForm: 'sourceSubForm'
+            },
```
![Screenshot from 2020-08-27 12-18-48](https://user-images.githubusercontent.com/9753817/91470151-593f9680-e862-11ea-8c7a-02939ed2536d.png)
![Screenshot from 2020-08-27 12-19-01](https://user-images.githubusercontent.com/9753817/91470153-59d82d00-e862-11ea-844b-0e2e736bb655.png)
![Screenshot from 2020-08-27 12-19-13](https://user-images.githubusercontent.com/9753817/91470154-59d82d00-e862-11ea-895f-06197eeb99c6.png)

cc @dsesami @AlanCoding @chrismeyersfsu

jlmitch5 reviewed (approved):
UI code looks good!

AlanCoding committed:
Get rid of ansible version checking

AlanCoding commented:
put up https://github.com/chrismeyersfsu/awx/pull/6

chrismeyersfsu committed:
remove backwords migraiton support for inv plugins

* Do not write out inventory source_vars to a file on disk as they _may_
contain sensitive information. This also removes support for backwards
migrations. This is fine, backwards migration is really only useful
during development.

chrismeyersfsu committed:
bump inv plugin migration to avoid conflict

chrismeyersfsu committed:
more get_ansible_version removal

chrismeyersfsu committed:
flake8 fix

chrismeyersfsu committed:
more get_ansible_version test failure fixes

unlikelyzero commented:
Please merge this at the same time:
https://github.com/ansible/tower-qa/pull/5464

Unrelated bugs discovered during testing:
https://github.com/ansible/awx/issues/8044
https://github.com/ansible/awx/issues/8045

AlanCoding commented on awx/main/tasks.py:
```diff
@@ -2567,23 +2562,18 @@ def build_args(self, inventory_update, private_data_dir, passwords):
+        else:
+            if getattr(settings, '%s_ENABLED_VAR' % src.upper(), False):
+                args.extend(['--enabled-var',
+                            getattr(settings, '%s_ENABLED_VAR' % src.upper())])
```
so now these become the defaults?

AlanCoding replied:
answer: yes

AlanCoding reviewed (approved):
From a technical perspective, I think got all the questions answered about expectations for how problems are solved. Really happy to see this happen 👍  it'll help a lot moving to jobs in containers

dsesami reviewed (approved):
approved from a UI perspective.

ryanpetrello commented:
recheck

ryanpetrello committed:
update the inventory source module to respect new fields
